Use docker in travis-ci for all tests.

cathay4t opened the pull request.

cathay4t commented:
Waiting on PR #232

cathay4t commented on automation/README.md:
```diff
@@ -1,36 +1,40 @@
+`./automation/run-tests.sh`
 
 ## Development
 
```
Done. Thanks.

cathay4t commented:
V2 patch set uploaded.

Changes since V1:
 * Patch reorganized by spliting tox patch out of travis-ci patch.
 * Add new patch to enable python 3.7 unit test.

tyll commented on automation/run-tests.sh:
```diff
@@ -68,17 +71,31 @@ function install_nmstate {
+        docker_exec 'cd /workspace/nmstate && tox -e check-py36'
+    fi
+
+    if [ $TEST_TYPE == "all" ] || [ $TEST_TYPE == "integ" ];then
```
The tox.ini also contains a definition to run integration tests for both python 2.7 and 3.6 - why are we defining how to run them differently here? Can we remove the definitions from tox or can we also update it and pass in the values for coverage in there somehow?

I just realized that for all, it will not execute all tests when an earlier one fails. IMHO it would be nice to run the integration tests first and ideally ensure that all tests run even with the earlier fail. If we can run the integration tests via tox, we could use just one tox call to run all tests, i.e. 'tox -e integration,flake8,pylint,check-py27, ...' to avoid early exits.

cathay4t replied:
That's a good point. Thanks. I will check whether we can fully depend on tox to kick off integration tests.

cathay4t replied:
When using tox, the nmstate is installed by pip. But we want to test the rpm, or should we both?

tyll replied:
[`skip_install=true`](https://tox.readthedocs.io/en/latest/config.html#conf-skip_install) should allow to avoid installing nmstate in the virtual env. For me it is enough to only test the RPM.

tyll commented on automation/run-tests.sh:
```diff
@@ -14,6 +14,9 @@ fi
 NET1="nmstate-net1"
 
+#Valid TEST_TYPE are: all, tox_code_style, tox_py27, tox_py36, integ
+: ${TEST_TYPE:="integ"}
```
@EdDev would you be fine with running all tests by default but allowing you to run integration tests only by setting this variable?

All would be both unit and integration tests.

EdDev replied:
I'm fine with that.

cathay4t replied:
Done.

cathay4t commented:
V3 patch set uploaded.

Changes since V2:
 * New patch to support upload code coverage data outside of container.
 * Expanded the test to run unit test on CentOS also.

Still waiting on PR #232

EdDev commented on automation/README.md:
```diff
@@ -27,16 +26,28 @@ It may be used both locally and through CI.
+By default, `./automation/run-tests.sh` will run integration test only in
+container. You may change the test type by using:
+
+ * `TEST_TYPE=unit_code_style ./automation/run-tests.sh`: Code style tests.
```
It will be nice to use a short name.
You can just use the term `lint` or `linter`.

And it is not limited to code stile, it finds bugs and missuses. I would name it `static analysis of code`.

cathay4t replied:
Done

EdDev commented on automation/run-tests.sh:
```diff
@@ -64,17 +67,31 @@ function install_nmstate {
-        tests/integration \
-    ${nmstate_pytest_extra_args}"
+    if [ $TEST_TYPE == "all" ] || [ $TEST_TYPE == "unit_code_style" ];then
+        docker_exec 'cd /workspace/nmstate && tox -e flake8,pylint'
```
We could just build the command in the "if/s" and execute it once after.

`docker_exec 'cd /workspace/nmstate && ` is duplicated all over.

Just a nice to have, not a must.

cathay4t replied:
Done.

EdDev commented on automation/run-tests.sh:
```diff
@@ -64,17 +67,31 @@ function install_nmstate {
-        --cov-report=html:htmlcov-integ \
-        tests/integration \
-    ${nmstate_pytest_extra_args}"
+    if [ $TEST_TYPE == "all" ] || [ $TEST_TYPE == "unit_code_style" ];then
```
Please define the names in constants.

cathay4t replied:
Done

EdDev commented on automation/run-tests.sh:
```diff
@@ -64,17 +67,31 @@ function install_nmstate {
+    fi
+
+    if [ $TEST_TYPE == "all" ] || [ $TEST_TYPE == "integ" ];then
+        docker_exec "
```
Please use a different function for the integration  and unit tests.

For the unit, you could pass the type as an argument.

tyll replied:
Why does it matter, whether or not they are integration or unit tests?

EdDev commented on .travis.yml:
```diff
@@ -1,11 +1,23 @@
-language: python
-python:
-  - 3.6
+sudo: required
```
Previously we ran inside a container, now we run on a VM.
It is important to understand that this may impact the resources we get from the CI. We should monitor this.

Please mention this detail in the commit message.

EdDev commented on .travis.yml:
```diff
@@ -1,11 +1,23 @@
+        - DOCKER_IMAGE=nmstate/fedora-nmstate-dev TEST_TYPE=integ
+        - DOCKER_IMAGE=nmstate/fedora-nmstate-dev TEST_TYPE=unit_code_style
+        - DOCKER_IMAGE=nmstate/fedora-nmstate-dev TEST_TYPE=unit_py27
+        - DOCKER_IMAGE=nmstate/fedora-nmstate-dev TEST_TYPE=unit_py36
```
What about py37?

EdDev replied:
ohh, found it in the next patch.

EdDev commented on .travis.yml:
```diff
@@ -1,11 +1,23 @@
-sudo: false
-dist: trusty
```
There is a new distribution: "Xenial" based on Ubunto 16.04.
Will that solve the problem of PyGobject dependency?

EdDev commented on automation/id_rsa_nmstate:
```diff
@@ -0,0 +1,51 @@
+-----BEGIN RSA PRIVATE KEY-----
```
How about using https://docs.travis-ci.com/user/encryption-keys ?

tyll replied:
It does not work for external PRs.

tyll commented on tox.ini:
```diff
@@ -78,6 +78,24 @@ commands =
+        --durations=5 \
+        --cov=libnmstate \
+        --cov=nmstatectl \
+        --cov-report=html:htmlcov-py37 \
```
Please also add `--cov-report=term \`

tyll replied:
my previous comment is not yet addressed.

cathay4t replied:
Done.

tyll commented on automation/upload_test_artifacts.sh:
```diff
@@ -0,0 +1,41 @@
+BASE_URL="http://grisge.info/nmstate"
+
+if [ -z "$SCP_KEY_SYMMETRIC_KEY" ];then
+    echo "SCP_KEY_SYMMETRIC_KEY is emoty, quiting"
```
Typo: `emoty` instead of `empty`.

cathay4t commented:
V4 patch set uploaded.

Changes since V3:

 * Upload the test artifacts outside of container, using travis-ci `after_failure` method.
 * Will not run unit test on CentOS as its docker image does not tox yet. Will do that in another PR.

tyll commented:
@cathay4t does this still depend on some other PR? If not, please remove the label.

cathay4t commented:
@tyll Flag removed. Ready for review.

tyll commented on automation/README.md:
```diff
@@ -27,16 +26,30 @@ It may be used both locally and through CI.
+By default, `./automation/run-tests.sh` will run all tests in CI container.
+You may change the test type by using:
+
+ * `TEST_TYPE=lint ./automation/run-tests.sh`: Static analysis of code.
```
There is now an example about how to use long-options for run-tests.sh, IMHO it makes sense to add a `--test-type` flag  as well. Fortunately, it is not too complicated.

cathay4t replied:
Done.

tyll commented on automation/upload_test_artifacts.sh:
```diff
@@ -0,0 +1,29 @@
+BASE_URL="http://grisge.info/nmstate"
+
+if [ -z "$TRAVIS_JOB_ID" ];then
+    echo "Need to be run in traivs-CI, quiting"
```
There is a typo: `traivis` instead of `travis`.

cathay4t replied:
Done.

cathay4t commented:
V5 patch set uploaded.

Changes since V4:
 * Allow define test type vai `./run-test.sh --test-type integ`.
 * New patch to all run test in Fedora image: `./run-test.sh --fedora`.
 * Apply the suggestions in feedback.

EdDev commented on .stdci.yml:
```diff
@@ -5,4 +5,4 @@ Architectures:
         host-distro: same
 script:
-    from-file: automation/run-integration-tests.sh
+    from-file: env TEST_TYPE=integ automation/run-tests.sh
```
Is this really working? I thought it only accepts files as values.

cathay4t replied:
Changed to use argument:
```
from-file: automation/run-tests.sh --test-type integ
```

EdDev commented on automation/Dockerfile.fedora:
```diff
@@ -15,7 +15,9 @@ RUN dnf -y install --setopt=install_weak_deps=False \
                    python3-pyyaml python3-jsonschema python3-setuptools \
                    NetworkManager-ovs openvswitch libibverbs python36 \
-                   python3-gobject-base dnsmasq radvd python3-tox python2 && \
+                   python3-gobject-base dnsmasq radvd python3-tox python2 \
```
Why do we have python2 stuff here?

cathay4t replied:
In  order to test `tox -e check-py27`.

EdDev commented on .travis.yml:
```diff
@@ -21,3 +21,13 @@ before_install:
     - sudo ./automation/run-tests.sh
+
+after_success:
+    - sudo pip install python-coveralls
```
We used `coveralls` in the tox, can we keep using the same?

Can we just do this using tox? Not critical, but it will look cleaner IMO.

tyll replied:
Also we are currently using `coveralls` and not `python-coveralls` in tox. Do we need to change the implementation?

cathay4t replied:
Changed to tox. Done.

cathay4t replied:
Using the tox for coveralls will need a long package list of ubuntu for compiling gobject as tox will try to install nmstate before run coveralls. Not sure why.

tyll replied:
[`skip_install=true`](https://tox.readthedocs.io/en/latest/config.html#conf-skip_install) should allow to avoid installing nmstate in the virtual env.

EdDev commented on .travis.yml:
```diff
@@ -1,18 +1,22 @@
-language: python
-python:
-  - 3.6
+language: c
```
Why did you removed then `dist`?

I think you need to use `generic`.

cathay4t replied:
Let me try just remove that line and see how it goes.

EdDev commented on .travis.yml:
```diff
@@ -1,18 +1,22 @@
-after_success:
-  - tox -e coveralls
+script:
+    - sudo ./automation/run-tests.sh
```
Why `sudo` is needed?

cathay4t replied:
Let me remove it and see how it goes.
It's copy for old travis config I guess travis now enforcing root privilege now.

tyll commented on automation/Dockerfile.fedora:
```diff
@@ -15,7 +15,9 @@ RUN dnf -y install --setopt=install_weak_deps=False \
                    rpm-build git python3-devel python3-dbus python3-coveralls \
                    python3-pyyaml python3-jsonschema python3-setuptools \
                    NetworkManager-ovs openvswitch libibverbs python36 \
-                   python3-gobject-base dnsmasq radvd python3-tox python2 && \
```
This changes needs to be a separate PR to update the container image.

cathay4t replied:
Done.

tyll commented on tox.ini:
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = flake8, pylint, check-{py27,py36}
+envlist = flake8, pylint, check-{py27,py36,py37}
```
Please move the python 3.7 test to a different PR. It would also be nice to test against Python 3.8.

cathay4t replied:
Done.

tyll commented on .travis.yml:
```diff
@@ -1,18 +1,40 @@
   apt:
     packages:
-    - libgirepository1.0-dev
+        - realpath
```
It would be great to sort the list of packages by name.

cathay4t replied:
Done.

EdDev commented on automation/id_rsa_nmstate:
```diff
@@ -0,0 +1,51 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Could you please rename this file to be more specific to its usage?
I feel uncomfortable having it, but I see there is no other options at the moment.

cathay4t replied:
Done.

tyll commented:
This obsoletes #274

cathay4t commented:
> Also we are currently using `coveralls` and not `python-coveralls` in tox. Do we need to change the implementation?
Done.

tyll commented on automation/run-tests.sh:
```diff
@@ -1,8 +1,7 @@
-: ${DOCKER_IMAGE:=nmstate/centos7-nmstate-dev}
 EXPORT_DIR="$PWD/exported-artifacts"
 CONT_EXPORT_DIR="/exported-artifacts"
 if [[ $DOCKER_IMAGE == *"fedora"* ]];then
```
This check now needs to happen after the docker image command-line option was parsed.

cathay4t replied:
This is handled at line 163 after the `getopt()`.

tyll replied:
Line 163 sets the `DOCKER_IMAGE` variable but line 7 already checks the value of this variable to determine the python site path command. I do not see how this works correctly.

tyll replied:
it seems to me that this works in Travis only because DOCKER_IMAGE is set in travis.yml

cathay4t replied:
Oh, I noticed that, will fix it. Thanks.

cathay4t replied:
Done.

tyll commented on automation/upload_test_artifacts.sh:
```diff
@@ -0,0 +1,29 @@
+SCP_BASEDIR="upload"
+SCP_USERNAME="nmstate"
+SCP_KEY="./automation/upload_test_artifacts.key"
+BASE_URL="http://grisge.info/nmstate"
```
Just wondering, can you setup letsencrypt for this? Not a blocker, though.

cathay4t replied:
Done. Through I need create a cron job for auto-renew.

tyll commented on automation/run-tests.sh:
```diff
@@ -58,32 +59,49 @@ function dump_network_info {
+            --log-date-format='%Y-%m-%d %H:%M:%S' \
+            --log-format='%(asctime)s %(filename)s:%(lineno)d %(levelname)s %(message)s' \
+            --durations=5 \
+            --cov=\$($PYTHON_SITE_PATH_CMD)/libnmstate \
```
Why do we need to the full path here instead of just `libnmstate`?

cathay4t replied:
The pytest is using rpm installed libnmstate instead of folder in source folder. If using the `libnmstate` only, the pytest coverage will report zero coverage.

tyll replied:
Thank you. I see. If we can assume that nmstate will be installed only for one python version in one running container, it seems that this can be simplified using a shell glob (does not need `PYTHON_SITE_PATH_CMD` anymore):

```
diff --git a/automation/run-tests.sh b/automation/run-tests.sh
index f6a1cf0..8164f12 100755
--- a/automation/run-tests.sh
+++ b/automation/run-tests.sh
@@ -95,8 +95,8 @@ function run_tests {
             --log-date-format='%Y-%m-%d %H:%M:%S' \
             --log-format='%(asctime)s %(filename)s:%(lineno)d %(levelname)s %(message)s' \
             --durations=5 \
-            --cov=\$($PYTHON_SITE_PATH_CMD)/libnmstate \
-            --cov=\$($PYTHON_SITE_PATH_CMD)/nmstatectl \
+            --cov /usr/lib/python*/site-packages/libnmstate \
+            --cov /usr/lib/python*/site-packages/nmstatectl \
             --cov-report=html:htmlcov-integ \
             --cov-report=term \
             tests/integration \
```
And this might also again allow to move this into tox.ini (there are currently only outdated integration test environments that should probably be removed).

cathay4t replied:
Good point. Yes, there is only one python version of nmstate been installed in all tests.
I have created issue #285 to trace this in another PR.

cathay4t commented on automation/README.md:
```diff
@@ -15,28 +15,42 @@ It may be used both locally and through CI.
+- run-tests.packages: Includes the packages needed by the oVirt CI (STDCI)
+  worker.
+
+- run-tests.environment.yaml: Instruct the oVirt CI (STDCI) only run
```
Done.

cathay4t commented on automation/README.md:
```diff
@@ -15,28 +15,42 @@ It may be used both locally and through CI.
+Assuming *docker* is installed on the host, just run:
+`./automation/run-tests.sh`
+
+By default, `./automation/run-tests.sh` will run all tests in CI container.
```
Done.

New patch now contains:

```
By default, `./automation/run-tests.sh` will run integration tests in the
container using 'nmstate/centos7-nmstate-dev' docker image.
You may change the test type by using:

 * `./automation/run-tests.sh --test-type lint --fedora`:
   Static analysis of code using 'nmstate/fedora-nmstate-dev' docker image.

 * `./automation/run-tests.sh --test-type unit_py27 --fedora`:
   Unit tests in Python 2.7 using 'nmstate/fedora-nmstate-dev' docker image.

 * `./automation/run-tests.sh --test-type unit_py36 --fedora`:
   Unit tests in Python 3.6 using 'nmstate/fedora-nmstate-dev' docker image.

 * `./automation/run-tests.sh --test-type unit_py37 --fedora`:
   Unit tests in Python 3.7 using 'nmstate/fedora-nmstate-dev' docker image.
 * `./automation/run-tests.sh --test-type integ`:
   Integration tests using 'nmstate/centos7-nmstate-dev' docker image.

 * `./automation/run-tests.sh --test-type integ --fedora`:
   Integration tests using 'nmstate/fedora-nmstate-dev' docker image.

 * `./automation/run-tests.sh --fedora`:
   All tests above using 'nmstate/fedora-nmstate-dev' docker image.
```

cathay4t commented:
Rebased to current master.

cathay4t commented:
ci test please.

EdDev commented on automation/README.md:
```diff
@@ -46,21 +69,21 @@ order to run the tests. Setting the environment variable `debug_exit_shell`
 `DOCKER_IMAGE` variable:
 
-`DOCKER_IMAGE=local/centos7-nmstate-dev debug_exit_shell=1 ./automation/run-integration-tests.sh`
+`DOCKER_IMAGE=local/fedora-nmstate-dev debug_exit_shell=1 ./automation/run-tests.sh`
```
I the fedora image is the default now, perhaps you should mention here centos7.

cathay4t replied:
Done. Thanks.

EdDev commented on automation/run-tests.sh:
```diff
@@ -0,0 +1,218 @@
+        ;;
+    --help)
+        echo -n "$0 [--pytest-args=...] [--help] [--debug-shell] "
+        echo -n "[--centos] "
```
You mean `--el7`, right?

cathay4t replied:
Thanks. Fixed.

EdDev commented on .travis.yml:
```diff
@@ -1,18 +1,23 @@
-dist: xenial
```
Why did you removed it?

cathay4t replied:
We don't care the version now as we are using docker now.

cathay4t replied:
Since we are using docker now, we don't care about which ubuntu version travis is using.

EdDev commented on .travis.yml:
```diff
@@ -1,18 +1,23 @@
+env:
+    matrix:
+        - TEST_TYPE=integ DOCKER_IMAGE=nmstate/centos7-nmstate-dev
+        - TEST_TYPE=integ DOCKER_IMAGE=nmstate/fedora-nmstate-dev
```
Does each combination runs on its own VM or on the same one? How much time it takes now compared to before?

cathay4t replied:
Each has its own VM.

cathay4t commented:
> Support unit test in CI container commit message:
> I think you should mention the file rename last and focus on the options to run the various tests first.
> Please also mention that the stdci continues to run the integration tests in centos7 (it is not clear, especially when 'run-tests.environment.yaml' specify it and not the command arguments).
Done.

cathay4t commented:
Rebased to current master.

tyll commented:
travis.yml needs to run `sudo modprobe openvswitch` or the run-tests.sh script when it is running in travis to fix the CI.

cathay4t commented:
@tyll Fixed. Thanks.

cathay4t commented:
New patch uploaded. Removed the CentOS 7 integration test from travis.

cathay4t commented:
New patch uploaded. CentOS 7 integration test is added back to travis and marked as optional.

EdDev commented:
ci test please

cathay4t commented:
Rebased to current master.
